w3-aggregation.md

```diff
@@ -0,0 +1,359 @@
+
+```json
+{
+  "ran": "bafy...invocation",
+  "out": {
+    "ok": {
+      "status": "queued"
+    }
+  },
+  "fx": {
+    "join": { "/": "bafy...dequeue" }
+  },
+  "meta": {},
+  "iss": "did:web:spade.storage",
+  "prf": []
+}
+```
+
+### `aggregate/get`
+
+A Storefront principal can invoke a capability to get state of an accepted aggregate.
+
+> `did:web:web3.storage` invokes capability from `did:web:spade.storage`
+
+```json
+{
+  "iss": "did:web:web3.storage",
+  "aud": "did:web:spade.storage",
+  "att": [{
+    "with": "did:web:web3.storage",
+    "can": "aggregate/get",
+    "nb": {
+      "link": "bafy...aggregate",
+    }
+  }],
+  "prf": [],
+  "sig": "..."
+}
+```
+
+Once this invocation is executed, a receipt is generated with the aggregate information:
+
+```json
+{
+  "ran": "bafy...get",
+  "out": {
+    "ok": {
+      "deals": [
+        {
+          "dealId": 111,
+          "storageProvider": "f07...",
+          "status": "Active",
+          "pieceCid": "bag...",
+          "dataCid": "bafy...",
+          "dataModelSelector": "Links/...",
+          "activation": "2023-04-13T01:58:00+00:00",
+          "expiration": "2024-09-05T01:58:00+00:00",
+          "created": "2023-04-11T17:57:30.522198+00:00",
+          "updated": "2024-04-12T03:42:26.928993+00:00"
+        }
+      ]
+    },
+  },
+  "fx": {
+    "fork": []
+  },
+  "meta": {},
+  "iss": "did:web:spade.storage",
+  "prf": []
+}
+```
+
+### `offer/review`
+
+When a broker receives an `aggregate/offer` invocation from a Storefront Principal, an [Effect](https://github.com/ucan-wg/invocation/#7-effect) for this submission is created with join task to be performed asynchronously.
+
+```json
+{
+  "iss": "did:web:spade.storage",
+  "aud": "did:web:web3.storage",
+  "att": [{
+    "with": "did:web:spade.storage",
+    "can": "offer/review",
+    "nb": {
+      "link": "bafy...aggregate",
+    }
+  }],
+  "prf": [],
+  "sig": "..."
+}
+```
+
+Once this invocation is executed, a receipt is generated with the status of the task updated. Accepted aggregate receipt will provide aggregate status info:
+
+```json
+{
+  "ran": "bafy...review",
+  "out": {
+    "ok": {
+       "status": "accepted",
+       "link": "bafy...aggregate"
+    }
+  },
+  "fx": {
+    "fork": []
+  },
+  "meta": {},
+  "iss": "did:web:spade.storage",
+  "prf": []
+}
+```
+
+If offered aggregate is invalid, details on failing commPs are also reported:
+
+```json
+{
+  "ran": "bafy...invocation",
+  "out": {
+    "error": {
+      "status": "rejected",
+      "link": "bafy...aggregate",
+      "cause": [{
+        "commP": "commP",
+        "reason": "reasonCode",
+      }],
+    },
+  },
+  "fx": {
+    "fork": []
+  },
+  "meta": {},
+  "iss": "did:web:spade.storage",
+  "prf": []
+}
+```
+
+### Schema
+
+```ipldsch
+type Aggregate =
+  { "status": "queued", link: Link } |
+  { "status": "accepted", link: Link } |
+  { "status": "rejected", link: Link }
```

Comment:
This is not a valid IPLD schema syntax, following is:

```suggestion
type Aggregate union {
  | AggregateInfo "queued"
  | AggregateInfo "status"
  | AggregateInfo "rejected"
} representation inline {
  discriminantKey "status"
}

type AggregateInfo struct {
  link: Link
}
```

Reply:
That said I think using [keyed representation](https://ipld.io/docs/schemas/features/representation-strategies/#union-keyed-representation) is more idiomatic so we may want to use following instead:

```suggestion
type Aggregate union {
  | Link "queued"
  | Link "status"
  | Link "rejected"
} representation keyed
```